### ISSUE TITLE
How to save (and load) the transformer model

### ISSUE THREAD

**Issue by nielsvandermolen** [tensorflow/text#678]
https://www.tensorflow.org/text/tutorials/transformer

I am creating my own transformer model based on the tutorial example from TensorFlow. To be able to make predictions with the model it has to be saved and loaded in the SavedModel TensorFlow format. My goal is to save the model in one notebook and load it in another to make predictions.

The main issue in saving the model is the call function:

```
def call(self, x, enc_output, training,
           look_ahead_mask, padding_mask):
```

The Keras model seems not to like this multi-input approach and expects only one inputs Tensor. I did manage to save it by setting the input tensors to a list but loading the model gave another error.

```could not find matching function to call loaded from the savedmodel```
<img width="912" alt="Screenshot 2021-08-09 at 15 57 08" src="https://user-images.githubusercontent.com/16288161/128719598-a94a0ee0-cc61-4f75-8738-3a558e1afbfe.png">

Live demo:
https://www.twitch.tv/videos/1112813845 Around 03:49

Is there a way forward to be able to save and load the model for predictions? Do I need something like a signature? In general what steps are needed to save a transformer model like this? Can it work with the custom masking input (look ahead_mask, padding_mask)

**Comment by nielsvandermolen** [comment]
It seems I found a solution by using one Tensor instead of a list of tensors

This is the call function for people that have a similar issue. It is a bit of custom because it uses an extra dimension for the data to use the embeddings.
Dimensions:
````
[IndexesOfTrainingData]
[Timestamps for positional encoding]
[IndexesOfOutputData (for the decoder]
```` 

So the shape of the input is ``[Batch, 3, Sequence Length]``

Input of the encoder is  ``[Batch, 2, Sequence Length]``
```
[IndexesOfTrainingData]
[Timediff for positional encoding]
```

Input of the decoder is  ``[Batch, 2, Sequence Length -1]``  -- Note the minus one because of the masking!
```
[IndexesOfOutputData (for the decoder]
[Timediff for positional encoding]
``` 

Call function:
```
  def call(self, x, training=False, enc_padding_mask=None,
           look_ahead_mask=None, dec_padding_mask=None):
    inp = tf.gather(x, [0,1], axis=1)
    tar = tf.gather(x, [2,1], axis=1)
    rangeInputs = tf.range(0, self.sequence_lenght-1) 
    tarInput = tf.gather(tar, rangeInputs, axis=2)
    emb_inp_output = self.inputEmbeddingLayer(inp)
    enc_output = self.encoder(emb_inp_output, training, enc_padding_mask)
    emb_tar_output = self.outputEmbeddingLayer(tarInput)
    dec_output, attention_weights = self.decoder(
    emb_tar_output, enc_output, training, look_ahead_mask, dec_padding_mask)
```

The main logic is to include all of the labeled data in the training set and do the masking with the gather function with a range and split the input and tar.

Following the normal inputs Keras Tensorflow structure will make the usage of the Transformer model much more usable so there might be a case to rewrite the tutorial to support saving and loading of the model.

**Comment by gsalcianu** [comment]
(1) Is there a way to reproduce your problem?  E.g., a Google Colab Notebook with a piece of code that highlights the problem?

I've run the colab from the tutorial you reference, https://www.tensorflow.org/text/tutorials/transformer, and it works fine, at least up to and including the 

tf.saved_model.save(translator, export_dir='translator')

and the subsequent

reloaded = tf.saved_model.load('translator')

I've tried watching the video you linked to, https://www.twitch.tv/videos/1112813845:
there is no info at 3:49 and the video has more than 4h.  We need something easier to allow us to understand your problem.

(2) Looking at your
https://user-images.githubusercontent.com/16288161/128719598-a94a0ee0-cc61-4f75-8738-3a558e1afbfe.png

I see an error message complaining that you pass a tuple of tensors, instead of a list of TensorSpec objects.  Maybe you can pass an appropriate signature, containing the expected list of TensorSpec objects.

**Comment by nielsvandermolen** [comment]
Hi, it seems to be fixed recently (The export code is new). This issue was a duplicate of:

https://github.com/tensorflow/tensorflow/issues/36328
